message_passing/message_handler_waiting_compensation_cleanup/workflows.py

```diff
@@ -0,0 +1,121 @@
+            self.workflow_exit_exception.set_result(exc)
+            await workflow.wait_condition(workflow.all_handlers_finished)
+            raise exc
+
+    @workflow.update
+    async def my_update(self, input: UpdateInput) -> str:
+        """
+        This update handler demonstrates how to handle the situation where the
+        main Workflow method exits prematurely. In that case we perform
+        compensation/cleanup, and fail the Update. The Update caller will get a
+        WorkflowUpdateFailedError.
+        """
+        # Coroutines must be wrapped in tasks in order to use workflow.wait.
+        update_task = asyncio.Task(self._my_update())
+        # 👉 Always use `workflow.wait` instead of `asyncio.wait` in Workflow
+        # code: asyncio's version is non-deterministic.
+        first_completed, _ = await workflow.wait(  # type: ignore
+            [update_task, self.workflow_exit_exception],
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+        # 👉 It's possible that the update completed and the workflow exited
+        # prematurely in the same tick of the event loop. If the Update has
+        # completed, return the Update result to the caller, whether or not the
+        # Workflow is exiting.
+        if (
+            update_task in first_completed
+            or input.on_premature_workflow_exit == OnWorkflowExitAction.CONTINUE
+        ):
+            return await update_task
+        else:
+            await self._my_update_compensation_and_cleanup()
+            raise exceptions.ApplicationError(
+                f"The update failed because the workflow run exited: {await self.workflow_exit_exception}"
+            )
+
+    async def _my_update(self) -> str:
+        """
+        This handler calls a slow activity, so
+
+        (1) In the case where the workflow finishes successfully, the worker
+            would get an UnfinishedUpdateHandlersWarning (TMPRL1102) if the main
+            workflow task didn't wait for it to finish.
+
+        (2) In the other cases (failure, cancellation, and continue-as-new), the
+            premature workflow exit will occur before the update is finished.
+        """
+        # Ignore: implementation detail specific to this sample
+        self._update_started = True
+
+        await workflow.execute_activity(
+            my_activity, start_to_close_timeout=timedelta(seconds=10)
+        )
+        return "update-result"
+
+    async def _my_update_compensation_and_cleanup(self):
+        workflow.logger.info(
+            "performing update handler compensation and cleanup operations"
+        )
+
+    async def _run(self, input: WorkflowInput) -> str:
+        # Ignore this method unless you are interested in the implementation
+        # details of this sample.
+
+        # Wait until handlers started, so that we are demonstrating that we wait for them to finish.
+        await workflow.wait_condition(lambda: getattr(self, "_update_started", False))
```

Comment:
```suggestion
        await workflow.wait_condition(lambda: self._update_started)
```

And would recommend initializing it to `False` in the constructor

Reply:
I think it should be in view and IMO we should encourage explicit attributes over string-based ones, even for control-based attributes like this one. But it's not a big deal.